### PR TITLE
[Foxy backport] Fixed memory leak in Buffer::waitForTransform (#281)

### PR DESCRIPTION
**Pull request by mdtoom**
* Make BufferCore callback handle maintenance internal.

That is, there is really no reason for external users to know
about TranformableCallbackHandle at all.  Instead, hide all of
the implementation detail about the handles within BufferCore
itself.  This gets simplifies the API, and allows us to ensure
that cleanup happens properly.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>

### CODE REVIEW

**Review by clalancette — CHANGES_REQUESTED** [review]
The issue in general with this patch is that it breaks API and ABI.  We don't generally allow that in released distros.

In order to get this in for Foxy, we are going to have to change this patch so that it doesn't break API and ABI.  Without looking into detail I'm not quite sure what that involves, but it means we can't remove methods (so we need to re-add `addTransformableCallback` and `addTransformableCallback`), and we can't change the signature of `addTransformableRequest` (though we could add a new overload).

**Comment by mdtoom** [comment]
What about only taking the https://github.com/ros2/geometry2/pull/281/commits/2302fe3d653dd3efe3662fe52a8ba136723aa517 of that MR and not the second one that breaks the api?
The first commit should alone also fixes the problem.
Updated this MR accordingly.

File: tf2_ros/include/tf2_ros/buffer.h [hunk 1]
```diff
@@ -261,7 +261,11 @@ namespace tf2_ros
     CreateTimerInterface::SharedPtr timer_interface_;
 
     /// \brief A map from active timers to BufferCore request handles
-    std::unordered_map<TimerHandle, tf2::TransformableRequestHandle> timer_to_request_map_;
+    typedef struct {
+      tf2::TransformableRequestHandle request_handle;
+      tf2::TransformableCallbackHandle callback_handle;
+    } TransformableHandles;
+    std::unordered_map<TimerHandle, TransformableHandles> timer_to_request_map_;
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/buffer.h]: So this almost works, but it potentially changes the ABI by changing this private member variable.

To get around this, we need to figure out a way to keep this member variable the same but add in the functionality we want.  One way to do this is to leave the map as it is, but then add a second map (outside of the class) that maps between a `TransformableRequestHandle` and a `TransformableCallbackHandle`.  Then we could do a lookup in both maps to find what we need.  Note that if we go this way, we also need to add in locking, as it could be accessed from multiple threads simultaneously.

That's just one idea.  There may be other ways to fix this in an ABI-compatible way.

**Review by clalancette — CHANGES_REQUESTED** [review]

File: tf2_ros/include/tf2_ros/buffer.h (hunk 1 above)
Comment (mdtoom) [reply tf2_ros/include/tf2_ros/buffer.h]: Ah, understood. I created a seperate map to do just that. The locking is already done by the `timer_to_request_map_mutex_` so we don't have to worry about that.

But why is it important that this variable is not changed? It is a private variable so no other class will be able to access it.

**Comment by mdtoom** [comment]
I cannot track down why the test fails, any thoughts?

File: tf2_ros/include/tf2_ros/buffer.h (hunk 1 above)
Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer.h]: > It is a private variable so no other class will be able to access it.

Yep, that's right.  Because of this, changing it would still be API compliant.

> But why is it important that this variable is not changed? 

But the problem is that changing the type could change the *ABI*, and we guarantee that in stable releases as well.  That's why we can't change it here.

File: tf2_ros/include/tf2_ros/buffer.h [hunk 2]
```diff
@@ -262,6 +262,7 @@ namespace tf2_ros
 
     /// \brief A map from active timers to BufferCore request handles
     std::unordered_map<TimerHandle, tf2::TransformableRequestHandle> timer_to_request_map_;
+    std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle> timer_to_callback_map_;
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/buffer.h]: Unfortunately, this still breaks ABI, by adding a new member to a public class.

One ugly way to fix this that I've used in the past is to put a global variable into the `buffer.cpp` file which is a map of objects to maps.  Then when you need to look this up, you first look up the correct map by the object pointer address, and then you can look up the `tf2::TransformableCallbackHandle` from that map.  This maintains ABI, while still preserving the information you want.  Note that since this is a global map, it will also need locking since multiple threads would need to be in here.

So to summarize, the variables you would need to add at the top of `buffer.cpp` would be something like:

```
std::mutex g_object_map_to_cb_handle_mutex;
std::map<void*, std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle>> g_object_map_to_cb_handle;
```

And then when you go to use this, you take the mutex, use "this" to look up the correct map, then use the TimerHandle in `this` to lookup the `TransformableCallbackHandle`.

You can see an example of this pattern in https://github.com/ros2/rclcpp/pull/1056

**Review by clalancette — CHANGES_REQUESTED** [review]

File: tf2_ros/include/tf2_ros/buffer.h (hunk 2 above)
Comment (mdtoom) [reply tf2_ros/include/tf2_ros/buffer.h]: Does adding a buffer destructor maintain ABI? If not, how do you propose I remove objects `g_object_map_to_cb_handle`?

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer.h]: Apparently adding one can change the ABI: https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html .  So that doesn't seem to be an option.

How about when we remove items from the map, we check to see if the map at that particular address is empty?  If it is, then we remove the map.  This will also require other parts of the code to check to see if the map exists (and create it if it doesn't), but should maintain ABI.

Comment (mdtoom) [reply tf2_ros/include/tf2_ros/buffer.h]: That is not what I mean. I think having the map all the time is fine but not removing mappings to destroyed Buffer objects introduces a new memory leak. In ros2/rclcpp#1056 the mapping from the object to the list is removed from `g_object_map_to_cb_handle` in the destructor but we can't do that here. How do you propose we remove mappings in this MR?

Comment (clalancette) [reply tf2_ros/include/tf2_ros/buffer.h]: I don't have a good idea off-hand.  The tools we have to work with are adding new (non-virtual) class methods, new free functions, and new global variables.  We'd have to do something with that combination of tools to make this work.

**Comment by SteveMacenski** [comment]
An FYI: its been externally validated to fix this issue https://github.com/ros-planning/navigation2/issues/1889#issuecomment-726762860. Given Foxy is LTS and this is an issue preventing TF2 from being used on a long-running product, I think its a critical bug that should be merged.

**Comment by clalancette** [comment]
> An FYI: its been externally validated to fix this issue [ros-planning/navigation2#1889 (comment)](https://github.com/ros-planning/navigation2/issues/1889#issuecomment-726762860). Given Foxy is LTS and this is an issue preventing TF2 from being used on a long-running product, I think its a critical bug that should be merged.

I don't disagree, but we haven't found an ABI compatible solution yet.

**Comment by mdtoom** [comment]
Talked to @marting87 and finally found a solution. Clearing the class from `g_object_map_to_cb_handle` when the callback handle is removed. The only time the memory leak is still present is when a `Buffer` object is removed before the `timerCallback` for an waiting transform is executed. I think this is acceptable, as this fix does also not remove the `cb_handle` from `transformable_callbacks_` in that case.

File: tf2_ros/src/buffer.cpp [hunk 3]
```diff
@@ -44,10 +44,22 @@
 #define ROS_INFO printf
 #define ROS_WARN printf
 
-
 namespace tf2_ros
 {
 
+// Added to backport: https://github.com/ros2/geometry2/pull/281
+std::mutex g_object_map_to_cb_handle_mutex;
+std::map<void*, std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle>> g_object_map_to_cb_handle;
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]: Nit: `#include <map>` and `#include <unordered_map>` for this.

File: tf2_ros/src/buffer.cpp [hunk 4]
```diff
@@ -44,10 +44,22 @@
...
+
+void deleteTransfromCallbackHandle(void *class_ptr, const TimerHandle &timer_handle)
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]:
```suggestion
void deleteTransformCallbackHandle(void *class_ptr, const TimerHandle &timer_handle)
```

File: tf2_ros/src/buffer.cpp [hunk 5]
```diff
@@ -215,8 +227,14 @@ Buffer::waitForTransform(const std::string& target_frame, const std::string& sou
...
             // The request handle was found, so a timeout has not occurred
-            this->timer_interface_->remove(it->first);
-            this->timer_to_request_map_.erase(it->first);
+            auto timer_handle = it->first;
+            this->timer_interface_->remove(timer_handle);
+            this->timer_to_request_map_.erase(timer_handle);
+            {
+              std::lock_guard<std::mutex> lock(g_object_map_to_cb_handle_mutex);
+              // Add a new default constructed callback function.
+              deleteTransfromCallbackHandle(this, timer_handle);
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]:
```suggestion
              deleteTransformCallbackHandle(this, timer_handle);
```

File: tf2_ros/src/buffer.cpp [hunk 6]
```diff
@@ -266,20 +296,26 @@ Buffer::timerCallback(const TimerHandle & timer_handle,
...
+    {
+      std::lock_guard<std::mutex> lock(g_object_map_to_cb_handle_mutex);
+      auto timer_and_callback_it = g_object_map_to_cb_handle[this].find(timer_handle);
+      timer_is_valid = (g_object_map_to_cb_handle[this].end() != timer_and_callback_it);
+
+      if (timer_is_valid) {
+        callback_handle = timer_and_callback_it->second;
+      }
+
+      deleteTransfromCallbackHandle(this, timer_handle);
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]:
```suggestion
      deleteTransformCallbackHandle(this, timer_handle);
```

File: tf2_ros/src/buffer.cpp (hunk 4 above)
Comment (clalancette) [reply tf2_ros/src/buffer.cpp]: Also, I think the class pointer we are passing in is always `Buffer`, so that first parameter can become `Buffer *class_ptr`.

File: tf2_ros/src/buffer.cpp (hunk 3 above)
Comment (clalancette) [reply tf2_ros/src/buffer.cpp]: Also, we should mark both of these as either `static` or in an anonymous namespace just to keep the symbols private.

Comment (clalancette) [reply tf2_ros/src/buffer.cpp]: And I believe we are always storing a `Buffer *` in here, so we can change the key of the map to `Buffer *`.

File: tf2_ros/src/buffer.cpp [hunk 7]
```diff
@@ -44,10 +44,22 @@
...
 namespace tf2_ros
 {
 
+// Added to backport: https://github.com/ros2/geometry2/pull/281
+std::mutex g_object_map_to_cb_handle_mutex;
+std::map<void*, std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle>> g_object_map_to_cb_handle;
+
+void deleteTransfromCallbackHandle(void *class_ptr, const TimerHandle &timer_handle)
+{
+  g_object_map_to_cb_handle[class_ptr].erase(timer_handle);
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]: Is it ever the case that this can be called without the class pointer key first having been added?  It may make sense to use `g_object_map_to_cb_handle.at(class_ptr)` here instead (it can throw an exception, but at least it won't corrupt memory then).

The same goes for all uses of `operator[]` below; it bears thinking about whether we should use `at` instead.

File: tf2_ros/src/buffer.cpp [hunk 8]
```diff
@@ -215,8 +227,14 @@ Buffer::waitForTransform(const std::string& target_frame, const std::string& sou
...
           if (request_handle == it->second) {
             // The request handle was found, so a timeout has not occurred
-            this->timer_interface_->remove(it->first);
-            this->timer_to_request_map_.erase(it->first);
+            auto timer_handle = it->first;
+            this->timer_interface_->remove(timer_handle);
+            this->timer_to_request_map_.erase(timer_handle);
+            {
+              std::lock_guard<std::mutex> lock(g_object_map_to_cb_handle_mutex);
+              // Add a new default constructed callback function.
```
Comment (clalancette) [inline tf2_ros/src/buffer.cpp]:
```suggestion
              // Remove the callback function.
```

**Review by clalancette — CHANGES_REQUESTED** [review]
First off, great work in finding this ABI compatible solution.

I've left a bunch of comments in here.  There are no major problems that I see, just mostly cleanup and making the code work a bit more robustly.

Thanks!

File: tf2_ros/src/buffer.cpp (hunk 7 above)
Comment (mdtoom) [reply tf2_ros/src/buffer.cpp]: Should not be possible, since all calls of delete are done after it has been created first. But I agree with you that `at` is an better idea. I will keep using `operator[]` at when inserting an element, or do you prefer `insert()`?

Comment (clalancette) [reply tf2_ros/src/buffer.cpp]: I think I *slightly* prefer `insert()`, but I really don't have a strong opinion.

**Review by clalancette — APPROVED** [review]
Looks good to me.  I'll run CI on it next.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13067/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8013/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10790/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13047/)

**Comment by clalancette** [comment]
All right, the Windows compiler warnings are known errors (fixed in Rolling by https://github.com/ros2/rviz/pull/625 and https://github.com/ros2/rcutils/pull/311, but not yet backported to Foxy).

I'm going to go ahead and merge this.  @jacobperron FYI.  Thanks again @mdtoom for all of the work backporting this.
